### PR TITLE
[Concurrency] Disallow passing global actor state inout

### PR DESCRIPTION
**Pull request by etcwilde**
Like with any other actors, properties protected by a global actor can't be passed inout to async functions on other actors or other async functions.

Edit: There is a portion of the patch that is mostly an NFC refactoring. This is kept in a separate commit. The juicy bit is in the second commit.

Resolves: rdar://72186232

### CODE REVIEW

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Comment by etcwilde** [comment]
@swift-ci clean smoke test Linux Platform

**Review by kavon — APPROVED** [review]
👍🏼 LGTM

File: test/Concurrency/actor_inout_isolation.swift [hunk 1]
```diff
@@ -131,3 +131,23 @@ extension TestActor {
...
   }
 }
+
+// Verify global actor protection
+
+@globalActor
+struct MyGlobalActor {
+  static let shared = TestActor()
+}
+
```
Comment (kavon) [inline test/Concurrency/actor_inout_isolation.swift:141]: Can we add a positive / OK test where one `@MyGlobalActor` passes state `inout` to another function `@MyGlobalActor` sync function? (I assume that we're allowing this)

**Review by kavon — APPROVED** [review]
Thought of one more case right after I hit Submit on the first approval :). Once you add that coverage then I think we're good!

File: test/Concurrency/actor_inout_isolation.swift (hunk 1 above)
Comment (etcwilde) [reply test/Concurrency/actor_inout_isolation.swift:141]: Can do!

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Comment by etcwilde** [comment]
Added a bunch of tests for where it's okay to pass the global actor state inout.

As discussed offline with Kavon, I'll be throwing up a new PR with a test case for a top-level call to a synchronous function:

```
@MyGlobalActor var foo = 10

func bar(_ foo: inout Int)  { }
bar(&foo)
```

That's an issue with the call context, not the fact that it's inout.

**Pull request merged**